### ISSUE TITLE
Need MPMD supporting for GPU to use pipeline parallelism training large scale model

### ISSUE THREAD

**Issue by MoFHeka** [tensorflow/tensorflow#62736]
### Issue type

Feature Request

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

tf 2.15

### Custom code

No

### Current behavior?

Nowadays pipeline parallelism has been implemented in PyTorch for a long time. It's very useful for training a CTR model with Embedding pipeline or training a large language model between two machine.

### Standalone code to reproduce the issue

```shell
Maybe a easy send/recv construction like tpu with xla?
```

**Comment by SuryanarayanaY** [comment]
Hi @MoFHeka ,

Tensorflow supports Data Parallelism only now. Model parallelism is yet to support fully for training. But with help of Dtensors we can achieve both data and model parallelism as per attached [tutorial](https://www.tensorflow.org/tutorials/distribute/dtensor_ml_tutorial).

As per my understanding of pipeline parallelism its hybrid approach of data and model parallelism.

**Comment by dathudeptrai** [comment]
@MoFHeka I always felt like the XLA and Tensorflow/Jax actually support many hidden features but never mention or write the document for it :)).

**Comment by MoFHeka** [comment]
@dathudeptrai I really agree with that. Large projects often lead to difficulties in project management.

**Comment by SuryanarayanaY** [comment]
> As tensorflow/compiler/xla/hlo/experimental/auto_sharding/auto_sharding.cc show, it seems that ZeRO stage 3 and pipeline parallel were already supported by XLA.

Hi @MoFHeka ,
 
I doubt it, correct me if I am wrong. I can see from Tf2.14v code [auto_sharding.cc](https://github.com/tensorflow/tensorflow/blob/v2.14.1/tensorflow/compiler/xla/hlo/experimental/auto_sharding/auto_sharding.cc) XLA supports SPMD which is for data parallelism which is supported by TF. Can you point exactly which part you are referring to that you feel that XLA supports Model parallelism or MPMD. This may help us to escalate the issue to SME and get confirmation. Thanks!

**Comment by MoFHeka** [comment]
> > As tensorflow/compiler/xla/hlo/experimental/auto_sharding/auto_sharding.cc show, it seems that ZeRO stage 3 and pipeline parallel were already supported by XLA.
> 
> Hi @MoFHeka ,
> 
> I doubt it, correct me if I am wrong. I can see from Tf2.14v code [auto_sharding.cc](https://github.com/tensorflow/tensorflow/blob/v2.14.1/tensorflow/compiler/xla/hlo/experimental/auto_sharding/auto_sharding.cc) XLA supports SPMD which is for data parallelism which is supported by TF. Can you point exactly which part you are referring to that you feel that XLA supports Model parallelism or MPMD. This may help us to escalate the issue to SME and get confirmation. Thanks!

@SuryanarayanaY Please check these comments. It said "This can result in a strategy similar to **ZeRO stage 3**. NOTE: The combination of this branch with **pipeline parallel** is not tested."
https://github.com/tensorflow/tensorflow/blob/99d80a9e254c9df7940b2902b14d15914dbbbcd9/tensorflow/compiler/xla/hlo/experimental/auto_sharding/auto_sharding.cc#L3249

**Comment by dathudeptrai** [comment]
@MoFHeka I suggest you use jax instead. SOmething like this or flash-attention, FP8 training, int8 training, ... all available in jax with support from XLA. TF also used XLA but kinda hard to custom.

**Comment by MoFHeka** [comment]
> @MoFHeka I suggest you use jax instead. SOmething like this or flash-attention, FP8 training, int8 training, ... all available in jax with support from XLA (natively). TF also used XLA but kinda hard to custom.

@dathudeptrai Unfortunately, JAX also doesn't support many features, such as sequence parallelism. And at the XLA level, even with JAX, many features are actually only supported by TPU. 
One more important thing, I can't train the CTR model with JAX, which lacks too many things.

**Comment by dathudeptrai** [comment]
@MoFHeka https://github.com/NVIDIA/TransformerEngine/pull/602

**Comment by dathudeptrai** [comment]
@MoFHeka Yeah. Generally speaking, coding in jax is harder than pytorch and a bit easier than TF. About low level customization, I think jax is better. Performance wise in my experiments showed that jax is better than pytorch :). Even deepspeed + Flash-attention-2 + Pytorch still not as good as jax :)).

**Comment by MoFHeka** [comment]
@dathudeptrai Unfortunately, due to the lack of sequence parallelism, the compute utilization of Alpa is lower than that of Megatron with the same tensor parallelism optimization. Because Alpa use too much device memory when using TP, which leads to a smaller batch size.

Besides, the CTR models really can't be trained with Jax. The Jax ecosystem of online services, data processing, and other components (such as Keras) is way too far behind TF.

**Comment by dathudeptrai** [comment]
@MoFHeka Why not use both TF and Jax at the same time :). Also please check out some advanced attention techniques recently introduced in jax (https://github.com/lhao499/large-sequence-modeling/tree/main).

**Comment by MoFHeka** [comment]
> @MoFHeka Why not use both TF and Jax at the same time :), you can call jax code in TF code nowadays. Also please check out some advanced attention techniques recently introduced in jax (https://github.com/lhao499/large-sequence-modeling/tree/main).

@dathudeptrai Yes, that's right, the Jax kernel can be used in TF code, although there’s no big difference between Jax kernels and Keras layers with XLA. 
But the problem is that the **pipeline parallelism** capabilities of JAX cannot be used in TF. TF currently lacks pipeline parallelism components.
 
Even in recent updates, DTensor is used to support tensor parallelism. But in the training of the **CTR model**, one of the biggest usage scenarios of TF, what is more needed is the ability of pipeline parallelism.

**Comment by MoFHeka** [comment]
@SuryanarayanaY Hi~? Is there any way to implement pipeline training in tensorflow?
'tf.distribute.experimental.rpc.Server' with 'server.register' looks like a good choice, but I'm not sure.

**Comment by MoFHeka** [comment]
@SuryanarayanaY Any progress? Tensorflow seems to be way behind in its competition with Pytorch...

**Comment by ptoulme-aws** [comment]
Jax supports both sequence and pipeline parallel. 
Pipeline parallel is just a scan loop over weights that are [PP_degree, weight1, weight2]
https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/scaling/JAX/pipeline_parallel_simple.html

Then for sequence parallel just bound your Norms or sequence parallel region in below sharding constraints

```
activation = with_sharding_constraint(activation, PartitionSpec(('fsdp', 'data'), 'model', None) # enter sequence parallel
some_compute.......
activation = with_sharding_constraint(activation, PartitionSpec(('fsdp', 'data'), None, None) # exit sequence parallel

**Comment by MoFHeka** [comment]
> Jax supports both sequence and pipeline parallel. Pipeline parallel is just a scan loop over weights that are [PP_degree, weight1, weight2] https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/scaling/JAX/pipeline_parallel_simple.html
> 
> Then for sequence parallel just bound your Norms or sequence parallel region in below sharding constraints
> 
> ```
> activation = with_sharding_constraint(activation, PartitionSpec(('fsdp', 'data'), 'model', None) # enter sequence parallel
> some_compute.......
> activation = with_sharding_constraint(activation, PartitionSpec(('fsdp', 'data'), None, None) # exit sequence parallel
> ```

@ptoulme-aws Yes, I know, but migrating Jax is not an easy task, such as it does not have a good inference platform. In addition, Jax does not support 1F1B pipeline parallel except enhanced by Ray .

**Comment by MoFHeka** [comment]
> It does support 1F1B. Check out Praxis library by Google or Axlearn by Apple.

@ptoulme-aws From what I known, Google doesn't seem to be planning to open source Praxis.

**Comment by ptoulme-aws** [comment]
It is open source - https://github.com/google/praxis

**Comment by MoFHeka** [comment]
> It is open source - https://github.com/google/praxis

@ptoulme-aws Sorry for my mistake. According to pathway's paper, it is not Praxis that is used in training 1F1B, but PLAQUE. For now, nobody answer this issue https://github.com/google/jax/issues/22438.

**Comment by fjmscut** [comment]
> @MoFHeka I always felt like the XLA and Tensorflow/Jax actually support many hidden features but never mention or write the document for it :)).

It is too difficult to use DP / PP / SP / Flash Attention 1 2 3 / ZeRO 1 2 3 in TensorFlow compare to PyTorch to train a llm at scale. It lags far behind pytorch. But I think with DTensor and other feature, it can actually be done. Why is no one maintaining the relevant documentation ...

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
